### PR TITLE
Build Bacon with Rollup from ES6 modules

### PR DESCRIPTION
**Pull request by lautis**
Converts the codebase to ES6 modules which are built with [Rollup](http://rollupjs.org) to produce browser/commonjs builds. Rollup gives us clean output (no requires in browser builds!) but also cleaner separation of scopes in source files.

While the end result isn't at all different, it's now easier to use JS tooling for Bacon source. Eslint detects unused non-prototype imports with `no-unused-vars` (not enabled, there are violations in code). The code can be loaded in node without separate compilation step – assuming "babel-register" is required.

Running individual test files is much faster and failures report line numbers in source files and not some random numbers in `dist/Bacon.js`. With some effort, it should be possible to even run the individual test file after file is modified since the mapping form source to test files is quite clean.

The benchmarks also run without requiring separate compilation step.

TODO:
- [x] easy interface to assemble partial builds
- ~~jsnext package~~
- [ ] replace browserify with rollup in browsertest

### CODE REVIEW

**Comment by lautis** [comment]
Looked into providing a jsnext main package. Two solutions came up to my mind: including JS source files in npm package or providing a compiled ES6 bundle file.

Source files would be a clean solution, but might cause some grievance in long term. Babel or ES6 can change, potentially breaking our code. We're not even really using the standard ES6 mode but loose mode for some transformation. Thus, it could also be something unexpected in Bacon code that is only exposed on native ES6 mode.

A separate bundle file with mostly ES5 but ES6 imports would solve these issues. However, that wouldn't be very useful since Bacon only provides one export. Essentially it would save the potentially 20 bytes or so, but also mean that we'd need to add yet another bundle file.

**Comment by lautis** [comment]
@raimohanska, @phadej, do you have any feedback?

**Comment by raimohanska** [comment]
Gotta admit I'm not so deep in the world of packaging javascript modules in so-many-different-ways :( The current build works pretty ok, I guess, and is a result of god-knows-how-many iterations of trial and error.

So I can only comment on this on a "project management level". My priotities are

1. Don't break it. Fortunately we have a great deal of tests that keep the functional things in check. Personally I rely on tests when making changes. The packaging / module system compatibility part is harder to test automatically though, I guess.
2. I (pretty much) insist that we still need have a trivial way to import the whole Bacon lib using npm and commonjs, as well as importing the lib from cdnjs using a script tag.
3. Keep/make baconjs a good citizen in all other relevant ecosystems / ways to build javascript software (there are probably ways I've never heard of)
4. Don't mess with performance
5. Don't break compatibility with dinosaur browsers (using shims is of course a possibility too)
6. Make it possible (one way or other) to get a subset of the library if you're into keeping things 
lightweight
7. Future proof / ES6+ compatibility whatever this means in practise

Damn it was hard to prioritize these thingsies. How do you feel about these?

**Comment by lautis** [comment]
That's a good list. This viikateisku the first rule a bit by introducing code churn but hipeille compensates with more standard roolina and faster test cycle. 

For library consumers, not much should change. The bundle is essentially same. Option for es6 module exists at least of you'd use Bacon from git.

**Comment by raimohanska** [comment]
@lautis wat? looks like you're using predictive typing with Finnish vocabulary :)

**Comment by lautis** [comment]
Yep, that's what happened. Fixed :blush:

File: assemble.js [hunk 1]
```diff
@@ -97,27 +83,50 @@ var main = function(options){
...
+    entry: 'src/bacon.js',
+    plugins: plugins
+  }).then((bundle) => {
+    return bundle.write({
+      format: 'umd',
+      moduleName: 'Bacon',
+      globals: {
+        jQuery: 'jQuery',
+        Bacon: 'Bacon',
+        Zeptop: 'Zepto'
```
Comment (alexej-d) [inline assemble.js]: It should be Zepto without "p" shouldn't it?

Comment (lautis) [inline assemble.js]: Yes, nice catch!

**Comment by raimohanska** [comment]
OMG this PR was totally forgotten. This might be a good thing for bacon.js 2.0 where some breaking changes are about to happen. Should we do this?

**Comment by darrennolan** [comment]
I think Rollup is a fantastic idea for 3rd party lib vendors.
Facebook's React moved it to after browserify.

The bundles produced tend to be far more absorbable/smaller.

Having said that, I haven't had much do with setting it up just yet, though we are investigating it atm.

**Comment by raimohanska** [comment]
I started rebasing this on the current master. If I succeed, I'll try it on the 2.0 branch...

**Comment by raimohanska** [comment]
Merged on top of current master, into new branch `rollup`. See #707

**Pull request closed without merging**